### ISSUE TITLE
Random handful of redacted message events visible after joining room with history_visibility==joined

### ISSUE THREAD

**Issue by richvdh** [matrix-org/synapse#1943]
`@​travis:t2l.io` joined a room over federation; having done so, he saw two "empty" events in the timeline from before he joined: http://imgur.com/ZC1k3IL.

It is expected that the events be empty - the remote federation server will have served redacted events because nobody on travis's homeserver was a member of the room in these periods. What is *not* expected is that travis's homeserver served them to the client over /messages.

Inspection of the database showed that the room state at each event was as expected. The problem went away after a synapse restart. This therefore appears to have been caused by a stale cache within synapse.

It is interesting to note that the events in question (`$148698101813672eDBdZ:chat.weho.st` and `$14877908446723tbDHY:chat.weho.st`) were the boundary points for the federation backfill requests:

```
2017-02-23 10:46:26,450 - synapse.http.outbound - 124 - INFO - GET-237401- {GET-O-85787} [chat.weho.st] Sending request: GET matrix://chat.weho.st/_matrix/federation/v1/backfill/!XMHzVqhLqEbFpqBFaf:chat.weho.st/?limit=100&v=%2414877908446723tbDHY%3Achat.weho.st
2017-02-23 10:46:55,305 - synapse.http.outbound - 124 - INFO - GET-237556- {GET-O-85791} [chat.weho.st] Sending request: GET matrix://chat.weho.st/_matrix/federation/v1/backfill/!XMHzVqhLqEbFpqBFaf:chat.weho.st/?limit=100&v=%24148698101813672eDBdZ%3Achat.weho.st
```

**Comment by turt2live** [comment]
I've managed to reproduce this (again) on Synapse 0.19.3-rc2 by joining `#freenode_#smartos:matrix.org`. I'm not too comfortable with dumping logs here, so feel free to message me (`@travis:t2l.io`) and I'll provide the log package (11mb compressed, 145mb expanded).

**Comment by Ralith** [comment]
As far as I can recall, this has happened every single time I've joined an IRC portal room on matrix.org. It might be relevant that I'm always joining via my own HS.

**Comment by PureTryOut** [comment]
I can reproduce this since having my own homeserver. Running Synapse 0.33.9 now, it happens every time I join a new IRC room.

**Comment by babolivier** [comment]
Reproduced it last evening:
![image](https://user-images.githubusercontent.com/5547783/53009426-63920080-3433-11e9-9717-ac25cede6008.png)

**Comment by erikjohnston** [comment]
This should be fixed in #4718

**Comment by MadLittleMods** [comment]
Closing as https://github.com/matrix-org/synapse/pull/4718 looks merged and I don't see any further activity here ⏩
